_posts/2021-02-11-tabnet/tabnet.Rmd

```diff
@@ -0,0 +1,411 @@
+
+Finally, we circle back to the promise of interpretability, demonstrating what is offered by `tabnet` and ending in a short discussion.
+
+# In the flow with TabNet
+
+As usual, we start by loading all required libraries. We also set a random seed, on the R as well as the `torch` sides. When model interpretation is part of your task, you will want to investigate the role of random initialization.
+
+```{r}
+library(torch)
+library(tabnet)
+library(tidyverse)
+library(tidymodels)
+library(finetune) # to use tuning functions from the new finetune package
+library(vip) # to plot feature importances
+
+set.seed(777)
+torch_manual_seed(777)
+
+```
+
+Next, we load the dataset.
+
+```{r}
+# download from https://archive.ics.uci.edu/ml/datasets/HIGGS
+higgs <- read_csv(
+  "HIGGS.csv",
+  col_names = c("class", "lepton_pT", "lepton_eta", "lepton_phi", "missing_energy_magnitude",
+                "missing_energy_phi", "jet_1_pt", "jet_1_eta", "jet_1_phi", "jet_1_b_tag",
+                "jet_2_pt", "jet_2_eta", "jet_2_phi", "jet_2_b_tag", "jet_3_pt", "jet_3_eta",
+                "jet_3_phi", "jet_3_b_tag", "jet_4_pt", "jet_4_eta", "jet_4_phi", "jet_4_b_tag",
+                "m_jj", "m_jjj", "m_lv", "m_jlv", "m_bb", "m_wbb", "m_wwbb"),
+  col_types = "fdddddddddddddddddddddddddddd"
+  )
+```
+
+What's this about? In high-energy physics, the search for new particles takes place at powerful particle accelerators, such as (and most prominently) CERN's [Large Hadron Collider](https://home.cern/science/accelerators/large-hadron-collider). In addition to actual experiments, simulation plays an important role. In simulations, "measurement" data are generated according to different underlying hypotheses, resulting in distributions that can be compared with each other. Given the likelihood of the simulated data, the goal then is to make inferences about the hypotheses.
+
+Resulting from just such a simulation, the above dataset (@higgs) explores what features could be measured assuming two different processes. In the first process, two gluons collide, and a heavy Higgs boson is produced; this is the signal process, the one we're interested in. In the second, the collision of the gluons results in a pair of top quarks -- this is the background process.
+
+Via different intermediaries, both processes result in the same end products -- so tracking these does not help. Instead, what the paper authors did was simulate kinematic features (momenta, specifically) of decay products, such as leptons (electrons and protons) and particle jets. In addition, they constructed a number of high-level features, features that presuppose domain knowledge. In their article, they showed that in contrast to other machine learning methods, deep neural networks did nearly as well when presented with the low-level features (the momenta) only as with just the high-level features alone.
```

Comment:
```suggestion
Through different intermediaries, both processes result in the same end products -- so tracking these does not help. Instead, what the paper authors did was simulate kinematic features (momenta, specifically) of decay products, such as leptons (electrons and protons) and particle jets. In addition, they constructed a number of high-level features, features that presuppose domain knowledge. In their article, they showed that, in contrast to other machine learning methods, deep neural networks did nearly as well when presented with the low-level features (the momenta) only as with just the high-level features alone.
```

Reply:
For that added comma: the rule is basically that you do a double offset (commas on both ends) if you can read the sentence without it: "…they showed that deep neural networks…” 🤓

Reply:
I didn't know that - thanks!